google-cloud-firestore/src/main/java/com/google/cloud/firestore/TransactionRunner.java

```diff
@@ -0,0 +1,279 @@
+   * @param userCallback The user provided transaction callback
+   * @param userCallbackExecutor The executor to run the user callback on
+   * @param numberOfAttempts The total number of attempts for this transaction
+   */
+  public TransactionRunner(
+      FirestoreImpl firestore,
+      Transaction.AsyncFunction<T> userCallback,
+      Executor userCallbackExecutor,
+      int numberOfAttempts) {
+    this.span = tracer.spanBuilder("CloudFirestore.Transaction").startSpan();
+    this.firestoreClient = firestore;
+    this.firestoreExecutor = firestore.getClient().getExecutor();
+    this.userCallback = userCallback;
+    this.attemptsRemaining = numberOfAttempts;
+    this.userCallbackExecutor = userCallbackExecutor;
+    this.backoffAlgorithm =
+        new ExponentialRetryAlgorithm(
+            firestore.getOptions().getRetrySettings(), CurrentMillisClock.getDefaultClock());
+    this.nextBackoffAttempt = backoffAlgorithm.createFirstAttempt();
+  }
+
+  ApiFuture<T> run() {
+    this.transaction = new Transaction(firestoreClient, this.transaction);
+
+    --attemptsRemaining;
+
+    span.addAnnotation(
+        "Start runTransaction",
+        ImmutableMap.of("attemptsRemaining", AttributeValue.longAttributeValue(attemptsRemaining)));
+
+    final SettableApiFuture<Void> backoff = SettableApiFuture.create();
+
+    // Add a backoff delay. At first, this is 0.
+    this.firestoreExecutor.schedule(
+        new Runnable() {
+          @Override
+          public void run() {
+            backoff.set(null);
+          }
+        },
+        nextBackoffAttempt.getRandomizedRetryDelay().toMillis(),
+        TimeUnit.MILLISECONDS);
+
+    nextBackoffAttempt = backoffAlgorithm.createNextAttempt(nextBackoffAttempt);
+
+    return ApiFutures.transformAsync(
+        backoff, new BackoffCallback(), MoreExecutors.directExecutor());
+  }
+
+  /**
+   * Invokes the user callback on the user callback executor and returns the user-provided result.
+   */
+  private SettableApiFuture<T> invokeUserCallback() {
+    final SettableApiFuture<T> callbackResult = SettableApiFuture.create();
+    userCallbackExecutor.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            ApiFutures.addCallback(
+                userCallback.updateCallback(transaction),
+                new ApiFutureCallback<T>() {
+                  @Override
+                  public void onFailure(Throwable t) {
+
+                    callbackResult.setException(t);
+                  }
+
+                  @Override
+                  public void onSuccess(T result) {
+                    callbackResult.set(result);
+                  }
+                },
+                MoreExecutors.directExecutor());
+          }
+        });
+    return callbackResult;
+  }
+
+  /** Determines whether the provided error is considered retryable. */
+  private boolean isRetryableTransactionError(ApiException exception) {
```

Comment:
Can we move this method down into `RestartTransactionCallback` since it's the only thing calling this method?

(nit: This method can also be made static)

Reply:
Moved, but it cannot be static in the inner class anymore.